✨ Visibility trigger support for querySelectorAll

micajuine-ho opened the pull request:
Task 2 of #26823

- `getAmpElement()` => `getAmpElementOrElements` 
   - If experiment is on, returns **unique elements** of `scopedQuerySelectorAll()` for given selector
        - Error if `selectionMethod` is provided
   - Otherwise, return `getElement()` of selector
- Also changed what promise gets pushed into the `unlistenPromises`

micajuine-ho commented on extensions/amp-analytics/0.1/events.js:
```diff
@@ -1508,29 +1508,35 @@ export class VisibilityTracker extends EventTracker {
+          .then(elements => {
+            for (let j = 0; j < elements.length; j++) {
+              unlistenPromises.push(
+                visibilityManagerPromise.then(
```
Does this seem correct?

zhouyx commented on extensions/amp-analytics/0.1/analytics-root.js:
```diff
@@ -279,6 +280,30 @@ export class AnalyticsRoot {
+    return this.ampdoc.whenReady().then(() => {
+      const results = [];
+      const foundElements = scopedQuerySelectorAll(
+        dev().assertElement(this.ampdoc.getBody()),
```
Could you confirm that if embed would get its own ampdoc? If not I think we still need to pass in the context

micajuine-ho replied:
Current implantation uses `this.getRoot().querySelector(selector)`, which I think is a better approach.

zhouyx commented on extensions/amp-analytics/0.1/analytics-root.js:
```diff
@@ -279,6 +280,30 @@ export class AnalyticsRoot {
+      );
+
+      // Length is not supported in all browsers
+      if (foundElements && foundElements.length) {
```
why don't we return `foundElements`?

micajuine-ho replied:
Oops, was previously doing a check there but it got removed. Thanks!

micajuine-ho replied:
there was a check there previously. thanks for the catch

zhouyx commented on extensions/amp-analytics/0.1/analytics-root.js:
```diff
@@ -288,19 +313,61 @@ export class AnalyticsRoot {
+      ).then(elements => {
+        const uniqueElements = [];
+        for (let i = 0; i < elements.length; i++) {
+          for (let j = 0; j < elements[i].length; j++) {
```
`getElementByScopedQuerySelectorAll_` returns an array of elements, why when would it be a nested array?

micajuine-ho replied:
followed up offline about how to check for unique elements

micajuine-ho commented:
PTAL @zhouyx

zhouyx commented on extensions/amp-analytics/0.1/analytics-root.js:
```diff
@@ -279,6 +279,31 @@ export class AnalyticsRoot {
+      }
+      const elements = [];
+      for (let i = 0; i < foundElements.length; i++) {
+        if (this.contains(foundElements[i])) {
```
`this.contains`?

micajuine-ho replied:
[Used in the original getAmpElements()](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/analytics-root.js#L273-L277), to make sure that the elements selected are descendants of the `analytics-root`: [defined here](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/0.1/analytics-root.js#L175)

zhouyx commented on extensions/amp-analytics/0.1/analytics-root.js:
```diff
@@ -287,18 +312,72 @@ export class AnalyticsRoot {
+   * @param {Array} opt_elementsReference reference to unique elements
+   * @return {!Promise<!Array<!AmpElement>>} Array of AMP elements corresponding to the selector if found.
+   */
+  getAmpElementOrElements(
```
I have an idea, let me know what do you think. We create a function
`getElements(context, selectors, selectionMethod)`.
The this function will always return an array of unique element selected.
And then in the `event.js`, we would call
```
getElements().then(elements => {
 for (let i=0; i<elements.length; i++) {
   registerForElement_(element[i])
 }
})
```

micajuine-ho replied:
Hmm, I tired doing that version, however, we when register the element's listener, the `getReadyPromise()` requires the element and the selector. So `getElements()` would be more complicated (retaining unique elements while returning some sort of mapping between selector and elements). 

Do you still think it's worth to go down that route?

micajuine-ho replied:
Talked offline and decided to change `getReadyPromise()` (since we will need to change it in the future too) to support the `getElements()` route.

micajuine-ho commented:
@zhouyx Ready for another round of review

zhouyx commented on extensions/amp-analytics/0.1/events.js:
```diff
@@ -1506,32 +1508,37 @@ export class VisibilityTracker extends EventTracker {
-            )
-            .then(element =>
+      const selectors = Array.isArray(selector)
+        ? selector.filter((val, index) => selectors.indexOf(val) === index)
```
I'm curious if this actually works. You are trying to use `selectors` before initialization? 
And I think we should definitely throw a user error if selector contains duplicate entry.

micajuine-ho replied:
Ok, I'll add verification here

zhouyx commented on extensions/amp-analytics/0.1/events.js:
```diff
@@ -1487,7 +1487,9 @@ export class VisibilityTracker extends EventTracker {
               visibilitySpec,
-              this.getReadyPromise(waitForSpec, selector),
+              this.getReadyPromise(
+                this.getWaitForSpecForRootSelector_(waitForSpec, selector)
```
What about inline the function? It's only called here once.

micajuine-ho replied:
Sg

zhouyx commented on extensions/amp-analytics/0.1/test/test-analytics-root.js:
```diff
@@ -333,25 +333,108 @@ describes.realWin('AmpdocAnalyticsRoot', {amp: 1}, env => {
-      return root.getAmpElement(body, '#child').then(element => {
-        expect(element).to.equal(child);
-      });
+      const elements = await root.getAmpElementOrElements(body, '#child');
```
I see `getAmpElement` is still used for non experiment case. Please keep the test coverage.
And I don't see `getAmpElementOrElements` declared anywhere. Please verify.

micajuine-ho replied:
Oops, didn't fix the tests because one of the node modules has been acting up. Will fix now.

zhouyx commented on extensions/amp-analytics/0.1/test/test-analytics-root.js:
```diff
@@ -333,25 +333,108 @@ describes.realWin('AmpdocAnalyticsRoot', {amp: 1}, env => {
+        }
+      });
+
+      it('should find elements by ID', async () => {
```
This is basically testing the browser provided `querySelectorAll` API. I don't think we need to test browser API.

micajuine-ho replied:
Sg, fixed tests to reflect this.

jridgewell commented on extensions/amp-analytics/0.1/analytics-root.js:
```diff
@@ -287,18 +311,71 @@ export class AnalyticsRoot {
+        for (let i = 0; i < elementsLists.length; i++) {
+          this.verifyAmpElements_(elementsLists[i], selectors[i]);
+          for (let j = 0; j < elementsLists[i].length; j++) {
+            if (uniqueElements.indexOf(elementsLists[i][j]) === -1) {
```
For small lists, this will be quick enough.

micajuine-ho replied:
@zhouyx I think the list will be relatively small. What do you think?

micajuine-ho commented on extensions/amp-analytics/0.1/analytics-root.js:
```diff
@@ -287,18 +311,71 @@ export class AnalyticsRoot {
+      );
+      const elementsListsPromise = [];
+      for (let i = 0; i < selectors.length; i++) {
+        elementsListsPromise.push(
```
Went with the first option

micajuine-ho commented:
@zhouyx Ready for another round of review. Ignore the parenthesis changes (bad merge). 

- Changed `getReadyPromise()`
- Condensed `getAmpElements()`'s work into `getElementsByQuerySelectorAll_()`
- Fixed tests (including tests for `getReadyPromise()`)

zhouyx commented on extensions/amp-analytics/0.1/test/test-analytics-root.js:
```diff
@@ -333,25 +333,67 @@ describes.realWin('AmpdocAnalyticsRoot', {amp: 1}, (env) => {
+        child.id = 'myId';
+        child2.id = 'myId';
+        child.classList.add('myClass');
+        child2.classList.add('myClass');
```
nit: could you add a child3 that's won't be selected, to make sure we're not select every elements.
Also please don't assign the same id to different element, I think class is enough.

zhouyx commented on extensions/amp-analytics/0.1/test/test-analytics-root.js:
```diff
@@ -333,25 +333,67 @@ describes.realWin('AmpdocAnalyticsRoot', {amp: 1}, (env) => {
       });
     });
+
+    describe('get amp elements', () => {
```
To confirm, what will be the API provided after all changes.
`analyticsRoot.getElements()` and `analyticsRoot.getElement()`
or a single `analyticsRoot.getElements()` or anything?
The design would affect existing tests.

micajuine-ho replied:
We will have `analyticsRoot.getElements()` (added in this pr) and `analyticsRoot.getElement()`, previous tests cover `getElement()`.

micajuine-ho replied:
No, `getElement()` is still used by `getAmpElement()` which is staying (non-experiment case).

zhouyx commented on extensions/amp-analytics/0.1/test/test-analytics-root.js:
```diff
@@ -333,25 +333,67 @@ describes.realWin('AmpdocAnalyticsRoot', {amp: 1}, (env) => {
+      });
+
+      it('should allow not-found element for AMP search', async () => {
+        expectAsyncConsoleError(/Element "#unknown" not found/, 1);
```
Could you please expand the test a bit, so it covers the case when one of the selector is invalid?

zhouyx commented on extensions/amp-analytics/0.1/test/test-analytics-root.js:
```diff
@@ -333,25 +333,67 @@ describes.realWin('AmpdocAnalyticsRoot', {amp: 1}, (env) => {
+      });
+
+      it('should fail if the found element is not AMP for AMP search', async () => {
+        expectAsyncConsoleError(/required to be an AMP element/, 1);
```
Could you please expand the test a bit, so it covers the case when one of the selected element is not AMP?

micajuine-ho replied:
Sorry, can you explain how this test doesn't cover that?

micajuine-ho replied:
sg

zhouyx commented on extensions/amp-analytics/0.1/test/test-analytics-root.js:
```diff
@@ -333,25 +333,67 @@ describes.realWin('AmpdocAnalyticsRoot', {amp: 1}, (env) => {
+          root.getAmpElements(body, ['#child'], null, true)
+        ).to.be.rejectedWith(/required to be an AMP element/);
+      });
+    });
```
A few other tests I can think of
1. selectionMethod not valid
2. duplicate element selected
3. duplicate item in `selectors` array
4. `:host` within the selectors array

micajuine-ho replied:
Thanks for the suggestions :D

zhouyx commented on extensions/amp-analytics/0.1/test/test-analytics-root.js:
```diff
@@ -641,6 +683,49 @@ describes.realWin(
+        child2.classList.add('i-amphtml-element');
+      });
+
+      it('should find all elements by selector', async () => {
```
could you try one thing, add another element to the embed host and make sure it won't be selected.

zhouyx replied:
Ping on this comment : )

micajuine-ho replied:
@zhouyx Do you mean add another element that is non-amp?

zhouyx replied:
I meant add element with the className to the env.parentWin and then verify that they don't get selected. 
https://github.com/ampproject/amphtml/blob/master/testing/describes.js#L805

micajuine-ho replied:
Got it! Thanks for the explanation. Added an element called parentChild that is attached to the the parentWin.

zhouyx replied:
Looks great! Thanks

zhouyx commented on extensions/amp-analytics/0.1/test/test-events.js:
```diff
@@ -2052,6 +2052,111 @@ describes.realWin('Events', {amp: 1}, (env) => {
           });
         });
+
+        it('should fire event for all elements of the selector', async () => {
```
what `events.js` does is take an array of element and register visibility triggers. I'd say we don't need test coverage for `fire event for all elements of the selector` and `fire event per selector`. 
A single `fire event for multiple elements` should work. What do you think?

micajuine-ho replied:
Yea I agree, I added it initially just to be diligent :)

zhouyx commented on extensions/amp-analytics/0.1/test/test-events.js:
```diff
@@ -2052,6 +2052,111 @@ describes.realWin('Events', {amp: 1}, (env) => {
+          });
+        });
+
+        it('should only fire one event for each element', async () => {
```
same here, as long as we can verify the element list only contains one element we should be good.

zhouyx commented on extensions/amp-analytics/0.1/test/test-events.js:
```diff
@@ -2126,64 +2231,47 @@ describes.realWin('Events', {amp: 1}, (env) => {
             .once();
           // Default case: waitFor is not specified, AMP element selected
-          const promise2 = tracker.getReadyPromise(undefined, selector, target);
+          const promise2 = tracker.getReadyPromise('ini-load', target);
```
this test seems so unnecessary now, let's remove it.

micajuine-ho replied:
Test was removed as per comment above

zhouyx replied:
I still see the test for default case, am i missing anything?

micajuine-ho replied:
Oops, I misunderstood your first comment. Removed the default test :)

zhouyx commented on extensions/amp-analytics/0.1/test/test-events.js:
```diff
@@ -2126,64 +2231,47 @@ describes.realWin('Events', {amp: 1}, (env) => {
-        expect(tracker.getReadyPromise('none', undefined, undefined)).to.be
-          .null;
-        expect(tracker.getReadyPromise('none', ':root', undefined)).to.be.null;
-        expect(tracker.getReadyPromise('none', selector, target)).to.be.null;
```
we can leave this one?

micajuine-ho replied:
We've changed `getReadyPromise()` to only take a `waitForSpec` that gets calculated inside `addTrigger()`. See `readyPromiseWaitForSpec`.

All three of these expectations translate to         `expect(tracker.getReadyPromise('none')).to.be.null;` in this new version.

zhouyx replied:
True. Sounds good.

zhouyx commented on test/manual/amp-analytics/amp-analytics-multi-selector.html:
```diff
@@ -72,6 +77,15 @@
+        data-vars-id='ad3'
+        data-ad-width=300
+        data-ad-height=250
+        class="myClass">
```
nit: You can add `.myClass` to another component to make sure multiple elements got selected.

micajuine-ho commented on extensions/amp-analytics/0.1/analytics-root.js:
```diff
@@ -287,22 +317,60 @@ export class AnalyticsRoot {
+        'Cannot have selectionMethod defined with an array selector.'
+      );
+      return this.getElementsByQuerySelectorAll_(selectors).then((elements) =>
+        elements.filter((element, index) => elements.indexOf(element) === index)
```
For a webpage with 30 amp-ad elements and no throttle:

- 0.024ms avg with O(n^2)
```
 elements.filter((element, index) => elements.indexOf(element) === index)
```

- 0.058ms avg with O(n)
```
for (let i = 0; i < elements.length; i++) {
   if (!elements[i].hasAttribute('special')) {
      elements[i].setAttribute('special', 'yes');
      returnElements.push(elements[i]);
   }
}
```

Not sure if there's. better way to do O(n)

micajuine-ho commented:
@zhouyx Ready for review

zhouyx commented on extensions/amp-analytics/0.1/test/test-analytics-root.js:
```diff
@@ -333,25 +333,106 @@ describes.realWin('AmpdocAnalyticsRoot', {amp: 1}, (env) => {
+      it('should remove duplicate selectors', async () => {
+        child.classList.add('myClass');
+        expect(
+          await root.getAmpElements(body, ['.myClass', '.myClass'], null, true)
```
The check for duplicate selectors is not within `rootAmpElements`, you may want to move the test to test-event.js?

micajuine-ho replied:
Good call.

zhouyx commented on extensions/amp-analytics/0.1/test/test-events.js:
```diff
@@ -2126,64 +2126,47 @@ describes.realWin('Events', {amp: 1}, (env) => {
-            ':root',
-            undefined
-          );
+          const promise1 = tracker.getReadyPromise('ini-load');
```
promise1 is exact the same to promise now, let's remove it.

micajuine-ho replied:
👍

zhouyx commented on extensions/amp-analytics/0.1/test/test-events.js:
```diff
@@ -2202,28 +2185,16 @@ describes.realWin('Events', {amp: 1}, (env) => {
-            ':root',
-            undefined
-          );
+          const promise1 = tracker.getReadyPromise('render-start');
```
ditto

zhouyx commented on test/manual/amp-analytics/amp-analytics-multi-selector.html:
```diff
@@ -72,6 +78,249 @@
+        data-ad-height=250
+        class="myClass">
+      </amp-ad>
+    <amp-ad id='ad29' width=300 height=250
```
haha why 29 ads 😄 
LGTM

micajuine-ho replied:
Random number of cmd+c & cmd+v LOL

micajuine-ho commented on extensions/amp-analytics/0.1/analytics-root.js:
```diff
@@ -287,22 +320,59 @@ export class AnalyticsRoot {
+   * @param {boolean=} opt_multiSelectorOn multi-selector expriment
+   * @return {!Promise<!Array<!AmpElement>>} Array of AMP elements corresponding to the selector if found.
+   */
+  getAmpElements(context, selectors, selectionMethod, opt_multiSelectorOn) {
```
Should be able to. Good point

micajuine-ho commented on extensions/amp-analytics/0.1/test/test-analytics-root.js:
```diff
@@ -641,6 +723,54 @@ describes.realWin(
     });
 
+    describe('get amp elements', () => {
+      beforeEach(() => {
+        child.classList.add('i-amphtml-element');
+        toggleExperiment(win, 'visibility-trigger-improvements', true);
```
@zhouyx  Any idea why this would not work for the embed case? I'm guessing it has to do with the win I'm passing in.

micajuine-ho replied:
Fix was to use parentRoot.ampdoc.win :)

micajuine-ho commented:
@zhouyx Can you approve again for the bundle size check?
